crates/holochain_sqlite/src/db/p2p.rs

```diff
@@ -0,0 +1,291 @@
+
+        Ok(stmt
+            .query_row(named_params! { ":agent": &agent.0 }, |r| {
+                let r = r.get_ref(0)?;
+                let r = r.as_blob()?;
+                let signed = AgentInfoSigned::try_from(r)
+                    .map_err(|e| rusqlite::Error::ToSqlConversionFailure(e.into()))?;
+                Ok(signed)
+            })
+            .optional()?)
+    }
+
+    fn p2p_list(&self) -> DatabaseResult<Vec<AgentInfoSigned>> {
+        use std::convert::TryFrom;
+
+        let mut stmt = self
+            .prepare(P2P_SELECT_ALL)
+            .map_err(|e| rusqlite::Error::ToSqlConversionFailure(e.into()))?;
+        let mut out = Vec::new();
+        for r in stmt.query_map([], |r| {
+            let r = r.get_ref(0)?;
+            let r = r.as_blob()?;
+            let signed = AgentInfoSigned::try_from(r)
+                .map_err(|e| rusqlite::Error::ToSqlConversionFailure(e.into()))?;
+
+            Ok(signed)
+        })? {
+            out.push(r?);
+        }
+        Ok(out)
+    }
+
+    fn p2p_gossip_query(
+        &self,
+        since_ms: u64,
+        until_ms: u64,
+        within_arc: DhtArc,
+    ) -> DatabaseResult<Vec<KitsuneAgent>> {
+        let mut stmt = self
+            .prepare(P2P_GOSSIP_QUERY)
+            .map_err(|e| rusqlite::Error::ToSqlConversionFailure(e.into()))?;
+
+        let (storage_start_1, storage_end_1, storage_start_2, storage_end_2) =
+            split_arc(&within_arc);
+
+        let mut out = Vec::new();
+        for r in stmt.query_map(
+            named_params! {
+                ":since_ms": clamp64(since_ms),
+                ":until_ms": clamp64(until_ms),
+                ":storage_start_1": storage_start_1,
+                ":storage_end_1": storage_end_1,
+                ":storage_start_2": storage_start_2,
+                ":storage_end_2": storage_end_2,
+            },
+            |r| {
+                let agent: Vec<u8> = r.get(0)?;
+                Ok(KitsuneAgent(agent))
+            },
+        )? {
+            out.push(r?);
+        }
+        Ok(out)
+    }
+
+    fn p2p_prune(&self) -> DatabaseResult<()> {
+        let now = std::time::SystemTime::now()
+            .duration_since(std::time::SystemTime::UNIX_EPOCH)
+            .unwrap()
+            .as_millis() as u64;
+
+        self.execute(P2P_PRUNE, named_params! { ":now": now })?;
+
+        Ok(())
+    }
+}
+
+/// Owned data dealing with a full p2p_store record.
+#[derive(Debug)]
+struct P2pRecord {
+    agent: KitsuneAgent,
+
+    // encoded binary
+    encoded: Vec<u8>,
+
+    // additional queryable fields
+    signed_at_ms: i64,
+    expires_at_ms: i64,
+    storage_center_loc: u32,
+
+    // generated fields
+    storage_start_1: Option<u32>,
+    storage_end_1: Option<u32>,
+    storage_start_2: Option<u32>,
+    storage_end_2: Option<u32>,
+}
+
+fn split_arc(arc: &DhtArc) -> (Option<u32>, Option<u32>, Option<u32>, Option<u32>) {
```

Comment:
Seems that this is the more semantic return type, if this is going to be used widely then might be worth changing.

```suggestion
fn split_arc(arc: &DhtArc) -> (Option<(u32, u32)>, Option<(u32, u32)>) {
```